### ISSUE TITLE
Code signature not valid for use in process

### ISSUE THREAD

**Issue by jtaxen** [Shougo/vimproc.vim#284]
I just tried to install vimproc using Vundle to use the package eagletmt/ghcmod-vim. It downloaded the repository and I went into the folder and ran make, which exited without errors. When I now try to open any *.hs files, I get the following error message:

```
Error detected while processing function <SNR>50_vp_dlopen[1]..<SNR>50_libcall:
line    1:
dlerror = "dlopen(/Users/<my_username>/.vim/bundle/vimproc.vim/lib/vimproc_mac.so, 0x0005): code signature in (/Users/<my_username>/.vim/bundle/vimproc.vim/lib/vimproc_mac.so) not valid for use in process: mapped file has no cdhash, completely unsigned? Code has to be at least
ad-hoc signed."
E364: Library call failed for "vp_dlopen()"
Error detected while processing function <SNR>11_LoadFTPlugin:
line   17:
E170: Missing :endfor
```

The software I used is:

os: macOS 10.14.1
vim: 8.0
make: GNU Make 3.81

What can I do to make this work?

**Comment by Shougo** [comment]
It seems MacOS X security issue.
I don't know the solution.

The PR is wellcome.

**Comment by jtaxen** [comment]
Ok ok, I'll see what I can do and come back if I manage to solve it

**Comment by kardolus** [comment]
@jtaxen were you able to resolve this issue? Ran into the same problem

**Comment by jtaxen** [comment]
For some reason I could not reproduce the error when I tried to fix it, but you could try to use ad-hoc signing:
````
codesign --force -s - </path/to/application>
````
From the documentation: "If identity [`-s`] is the single letter "-" (dash), ad-hoc signing is performed.  Ad-hoc signing does not use an identity at all, and identifies exactly one instance of code."

Since I don't know enough about code signing conventions, I cannot tell if this is a good solution and since I could not reproduce the error I have not been able to test it throughoutly, but please let me know if you try it and it works, and I could maybe add a line in the README file under the build for Mac instructions.
See also this answer on Stackexchange: https://apple.stackexchange.com/questions/343912/should-i-sign-open-source-code-myself

**Comment by apocalypse2012** [comment]
I am getting this same error in Spacevim. I tried the adhoc code signing above with no success. Just more error puke...

Error detected while processing function SpaceVim#layers#debug#launching[2]..vebugger#pdb#start[5]..vebugger#std#startDebugger[1]..vebugger#startDebugger[3]..vebugger#createDebugger[4]..vimproc#ptyopen[6]..<SNR>222_plineopen[60]..<SNR>222_vp_pty_open[1]..<SNR>222_libcall:
line    1:
dlerror = "dlopen(/Users/raymond/.cache/vimfiles/repos/github.com/Shougo/vimproc.vim/lib/vimproc_mac.so, 0x0005): code signature in (/Users/raymond/.cache/vimfiles/repos/github.com/Shougo/vimproc.vim/lib/vimproc_mac.so) not valid for use in process: mapped file has no Team 
ID and is not a platform binary (signed with custom identity or adhoc?)"
E364: Library call failed for "vp_pty_open()"
Error detected while processing function SpaceVim#layers#debug#launching[2]..vebugger#pdb#start[5]..vebugger#std#startDebugger[1]..vebugger#startDebugger[3]..vebugger#createDebugger[4]..vimproc#ptyopen[6]..<SNR>222_plineopen[60]..<SNR>222_vp_pty_open:
line    1:
E714: List required
Error detected while processing function SpaceVim#layers#debug#launching[2]..vebugger#pdb#start[5]..vebugger#std#startDebugger[1]..vebugger#startDebugger[3]..vebugger#createDebugger:
line    9:
E121: Undefined variable: stdout
E15: Invalid expression: { 'out': {'pipe':(l:debugger.shell.stdout), 'buffer': ''}, 'err': {'pipe':(l:debugger.shell.stderr), 'buffer': '', 'annotation': "err:\t\t"}}
line   12:
E716: Key not present in Dictionary: pipes)
E116: Invalid arguments for function values(l:debugger.pipes)
E15: Invalid expression: values(l:debugger.pipes)
Error detected while processing function SpaceVim#layers#debug#launching[2]..vebugger#pdb#start[5]..vebugger#std#startDebugger[3]..vebugger#std#standardInit[1]..vebugger#std#setStandardBufferers:
line    1:
E716: Key not present in Dictionary: pipes)
E116: Invalid arguments for function values(a:debugger.pipes)
E15: Invalid expression: values(a:debugger.pipes)
Press ENTER or type command to continue
Error detected while processing function <SNR>244_readingTimerCallback[1]..205:
line    2:
E716: Key not present in Dictionary: pipes)

**Comment by Shougo** [comment]
Unfortunately, I don't have macOS environment.
You can fix the problem.

**Comment by alistar79** [comment]
Looks like you need to create a self signed cert under system and sign the binary yourself.

See https://gist.github.com/hlissner/898b7dfc0a3b63824a70e15cd0180154

Sadly I can't do this as the Mac I'm on is corporate controlled so if someone who has a mac developer account can do this we can all share that binary I think?

**Comment by thomasmarsh** [comment]
I hit this as well, and it took me a little bit of looking into Catalina and SIP to understand what is going on. This happens when you use the stock vim which is installed in /usr/bin. This directory is protected by SIP. The issue is that any protected and signed binary (such as /usr/bin/vim) cannot use dlopen to load unsigned libraries (such as vimproc_mac.so).

There are a couple solutions. One is to create a self-signed cert to sign the shared library yourself. This is rather painful. An even more extreme measure is to disable SIP.

A simpler solution, for me, is to reinstall vim (or nvim) in a non-SIP location, for example in /usr/local/bin. I did so with homebrew (`brew install nvim`). I set up an alias to make sure when I type `vi` or `vim`, that I'm using the version in /usr/local/bin.
